core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala

```diff
@@ -0,0 +1,166 @@
+        case Some(s) =>
+          s.receiveB(b)
+        case None =>
+          bCancel()
+      }
+    }
+
+    ret
+  }
+
+  /**
+    * Merges two streams with the second stream included asynchronously, with
+    * a value published from the resulting stream whenever this stream
+    * publishes a value. The periodicity of the resulting stream matches
+    * the periodicity of this stream.
+    * @param other the stream to merge with
+    * @tparam O the type of values from the other stream
+    * @return a stream with the values from both streams brought together
+    */
+  def zipAsync[O](other: Stream[O]): Stream[(T, O)] = {
+    val ret = new AsyncZippedStream[T, O](expectedPeriodicity)
+    val ptr = WeakReference(ret)
+
+    var aCancel: Cancel = null
+    aCancel = this.foreach { a =>
+      ptr.get match {
+        case Some(s) =>
+          s.receiveMaster(a)
+        case None =>
+          aCancel()
+      }
+    }
+
+    var bCancel: Cancel = null
+
+    bCancel = other.foreach { b =>
+      ptr.get match {
+        case Some(s) =>
+          s.receiveFollower(b)
+        case None =>
+          bCancel()
+      }
+    }
+
+    ret
+  }
+
+  /**
+    * Merges two streams with the the resulting stream being updated whenever either
+    * of its parents produce a new value. The resulting stream __does not__ have
+    * a periodic nature.
+    * @param other the stream to merge with
+    * @tparam O the type of values from the other stream
+    * @return a stream with the values from both streams brought together
+    */
+  def zipEager[O](other: Stream[O]): Stream[(T, O)] = {
+    val ret = new EagerZippedStream[T, O]
+    val ptr = WeakReference(ret)
+
+    var aCancel: Cancel = null
+    aCancel = this.foreach { a =>
+      ptr.get match {
+        case Some(s) =>
+          s.receiveA(a)
+        case None =>
+          aCancel()
+      }
+    }
+
+    var bCancel: Cancel = null
+
+    bCancel = other.foreach { b =>
+      ptr.get match {
+        case Some(s) =>
+          s.receiveB(b)
+        case None =>
+          bCancel()
+      }
+    }
+
+    ret
+  }
+
+  /**
+    * Adds a listener for elements of this Signal. Callbacks will be executed
+    * whenever a new value is published in order of when the callbacks were added.
+    * Callbacks added first will be called first and callbacks added last will
+    * be called last.
+    *
+    * @param thunk the listener to be called on each published value
+    * @return a function to call to remove the listener
+    */
+  def foreach(thunk: T => Unit): Cancel = {
+    listeners.enqueue(thunk)
+
+    () => {
+      listeners.dequeueFirst(_ eq thunk)
+    }
+  }
+}
+
+object Stream {
+  def periodic[T](period: Time)(value: => T)(implicit clock: Clock): Stream[T] = {
+    new Stream[T] {
+      val expectedPeriodicity = Periodic(period)
+
+      clock(period) { _ =>
+        publishValue(value)
+      }
+    }
+  }
+
+  def manual[T]: (Stream[T], T => Unit) = {
```

Comment:
The point is that you should only be able to manually publish values to the top-level stream created by `manual`, otherwise it would be possible to randomly publish values to intermediate streams created by transformations.